File: runtime/cs/bintalk/ProtocolReader.cs
using System;
using System.Text;
using System.Collections.Generic;

namespace bintalk
{
    /** This class can read basic types by using a bintalk.IReader object. */
    public static partial class ProtocolReader
    {
        public static bool readMid(bintalk.IReader r, ref ushort v)
        {
            return read(r, ref v, 0);
        }
		// int64
    	public static bool read(bintalk.IReader r, ref long v, uint maxValue)
    	{
            byte[] data; int startId;
            if (!r.read(8, out data, out startId)) return false;
            v = BitConverter.ToInt64(data, startId);
            return true;
    	}
		// uint64
    	public static bool read(bintalk.IReader r, ref ulong v, uint maxValue)
    	{
            byte[] data; int startId;
            if (!r.read(8, out data, out startId)) return false;
            v = BitConverter.ToUInt64(data, startId);
            return true;
    	}
		// double
    	public static bool read(bintalk.IReader r, ref double v, uint maxValue)
    	{
            byte[] data; int startId;
            if (!r.read(8, out data, out startId)) return false;
            v = BitConverter.ToDouble(data, startId);
            return true;
    	}
		// float
    	public static bool read(bintalk.IReader r, ref float v, uint maxValue)
    	{
            byte[] data; int startId;
            if (!r.read(4, out data, out startId)) return false;
            v = BitConverter.ToSingle(data, startId);
            return true;
    	}
		// int32
    	public static bool read(bintalk.IReader r, ref int v, uint maxValue)
    	{
            byte[] data; int startId;
            if (!r.read(4, out data, out startId)) return false;
            v = BitConverter.ToInt32(data, startId);
            return true;
    	}
		// uint32
    	public static bool read(bintalk.IReader r, ref uint v, uint maxValue)
    	{
            byte[] data; int startId;
            if (!r.read(4, out data, out startId)) return false;
            v = BitConverter.ToUInt32(data, startId);
            return true;
    	}
		// int16
    	public static bool read(bintalk.IReader r, ref short v, uint maxValue)
    	{
            byte[] data; int startId;
            if (!r.read(2, out data, out startId)) return false;
            v = BitConverter.ToInt16(data, startId);
            return true;
    	}
		// uint16
    	public static bool read(bintalk.IReader r, ref ushort v, uint maxValue)
    	{
            byte[] data; int startId;
            if (!r.read(2, out data, out startId)) return false;
            v = BitConverter.ToUInt16(data, startId);
            return true;
    	}
		// int8
    	public static bool read(bintalk.IReader r, ref sbyte v, uint maxValue)
    	{
            byte[] data; int startId;
            if (!r.read(1, out data, out startId))
                return false;
            v = (sbyte)data[startId];
            return true;
    	}
		// uint8
    	public static bool read(bintalk.IReader r, ref byte v, uint maxValue)
    	{
            byte[] data; int startId;
            if (!r.read(1, out data, out startId))
                return false;
            v = data[startId];
            return true;
    	}
		// bool
    	public static bool read(bintalk.IReader r, ref bool v, uint maxValue)
    	{
            byte[] data; int startId;
            if (!r.read(1, out data, out startId))
                return false;
            v = (data[startId] == 0)?false:true;
            return true;
    	}
        // string.
    	public static bool read(bintalk.IReader r, ref string v, uint maxValue)
    	{
            uint s;
            if (!readDynSize(r, out s) || s > maxValue)
                return false;
            byte[] data; int startId;
            if(!r.read(s, out data, out startId))
                return false;
            v = Encoding.UTF8.GetString(data, startId, (int)s);
            return true;
    	}
        // binary.
    	public static bool read(bintalk.IReader r, ref byte[] v, uint maxValue)
    	{
            uint s;
            if (!readDynSize(r, out s) || s > maxValue)
                return false;
            v = new byte[s];
            byte[] data; int startId;
            if(!r.read(s, out data, out startId))
                return false;
            Array.Copy(data, startId, v, 0, s);
            return true;
    	}
        // dynamic size.
        public static bool readDynSize(bintalk.IReader r, out uint s)
        {
            s = 0;
            byte b = 0;
            if (!read(r, ref b, 0))
                return false;
            uint n = (uint)((b & 0XC0) >> 6);
            s = (uint)(b & 0X3F);
            for (int i = 0; i < n; i++)
            {
                if (!read(r, ref b, 0))
                    return false;
                s = (s << 8) | b;
            }
            return true;
        }
    }
}

File: runtime/cpp/bintalk/ProtocolReader.cpp
#include "Config.h"
#include "ProtocolReader.h"

namespace bintalk
{

bool ProtocolReader::readMID(BinaryReader* r, MID& mid)
{
#if BINTALK_BIG_ENDIAN
	if(!r->read(&mid, sizeof(MID))) return false;
	mid = swapEndian(mid);
	return true;
#endif //BINTALK_BIG_ENDIAN
	return r->read(&mid, sizeof(MID));
}

#if BINTALK_BIG_ENDIAN
	#define BINTALK_PROTOCOLREADER_SINGLE_IMP(T)\
		bool ProtocolReader::read##T(BinaryReader* r, T& v, UINT32 maxArray, UINT32 maxValue)\
		{if(!r->read(&v, sizeof(T))) return false;\
		v = swapEndian<T>(v); return true;\ }
#else //BINTALK_BIG_ENDIAN
	#define BINTALK_PROTOCOLREADER_SINGLE_IMP(T)\
		bool ProtocolReader::read##T(BinaryReader* r, T& v, UINT32 maxArray, UINT32 maxValue)\
		{ return r->read(&v, sizeof(T)); }
#endif //BINTALK_BIG_ENDIAN

#define BINTALK_PROTOCOLREADER_ARRAY_IMP(T)\
bool ProtocolReader::read##T##A(BinaryReader* r, std::vector<T>& v, UINT32 maxArray, UINT32 maxValue)\
{\
	UINT32 s;\
	if(!readDynSize(r, s) || s > maxArray) return false;\
	v.resize(s);\
	for(UINT32 i = 0; i < s; i++) if(!read##T(r, v[i], maxArray, maxValue)) return false;\
	return true;\
}

#define BINTALK_PROTOCOLREADER_SIMPLE_IMP(T)\
	BINTALK_PROTOCOLREADER_SINGLE_IMP(T);\
	BINTALK_PROTOCOLREADER_ARRAY_IMP(T);

BINTALK_PROTOCOLREADER_SIMPLE_IMP(INT64);
BINTALK_PROTOCOLREADER_SIMPLE_IMP(UINT64);
BINTALK_PROTOCOLREADER_SIMPLE_IMP(DOUBLE);
BINTALK_PROTOCOLREADER_SIMPLE_IMP(FLOAT);
BINTALK_PROTOCOLREADER_SIMPLE_IMP(INT32);
BINTALK_PROTOCOLREADER_SIMPLE_IMP(UINT32);
BINTALK_PROTOCOLREADER_SIMPLE_IMP(INT16);
BINTALK_PROTOCOLREADER_SIMPLE_IMP(UINT16);
BINTALK_PROTOCOLREADER_SIMPLE_IMP(INT8);
BINTALK_PROTOCOLREADER_SIMPLE_IMP(UINT8);

bool ProtocolReader::readBOOL(BinaryReader* r, BOOL& v, UINT32 maxArray, UINT32 maxValue)
{
	UINT8 v1 = v?1:0;
	return r->read(&v, sizeof(BOOL));
}

bool ProtocolReader::readBOOLA(BinaryReader* r, std::vector<BOOL>& v, UINT32 maxArray, UINT32 maxValue)
{
	UINT32 s;
	if(!readDynSize(r, s) || s > maxArray) return false;
	v.resize(s);
	for(UINT32 i = 0; i < s; i++) 
	{
		BOOL b;
		if(!readBOOL(r, b, 0, maxValue)) 
			return false;
		v[i] = b;
	}
	return true;
}

bool ProtocolReader::readSTRING(BinaryReader* r, STRING& v, UINT32 maxArray, UINT32 maxValue)
{
	UINT32 len;
	if(!readDynSize(r, len) || len > maxValue)
		return false;
	v.resize(len);
	return r->read((void*)v.c_str(), len);
}
BINTALK_PROTOCOLREADER_ARRAY_IMP(STRING);

bool ProtocolReader::readBINARY(BinaryReader* r, BINARY& v, UINT32 maxArray, UINT32 maxValue)
{
	UINT32 len;
	if(!readDynSize(r, len) || len > maxValue)
		return false;
	v.resize(len);
	if(len > 0)
		return r->read((void*)(&(v[0])), len);
	return true;
}
BINTALK_PROTOCOLREADER_ARRAY_IMP(BINARY);

bool ProtocolReader::readDynSize(BinaryReader* r, UINT32& s)
{
	s = 0;
	UINT8 b;
	if(!r->read(&b, sizeof(UINT8)))
		return false;
	size_t n = (b & 0XC0)>>6;
	s = (b & 0X3F);
	for(size_t i = 0; i < n; i++)
	{
		if(!r->read(&b, sizeof(UINT8)))
			return false;
		s = (s<<8)|b;
	}
	return true;
}

}